Sources/System/IORing.swift

```diff
@@ -0,0 +1,657 @@
+@_implementationOnly import CSystem
+import Glibc  // needed for mmap
+import Synchronization
+
+@_implementationOnly import struct CSystem.io_uring_sqe
+
+// XXX: this *really* shouldn't be here. oh well.
+extension UnsafeMutableRawPointer {
+    func advanced(by offset: UInt32) -> UnsafeMutableRawPointer {
+        return advanced(by: Int(offset))
+    }
+}
+
+extension UnsafeMutableRawBufferPointer {
+    func to_iovec() -> iovec {
+        iovec(iov_base: baseAddress, iov_len: count)
+    }
+}
+
+// all pointers in this struct reference kernel-visible memory
+@usableFromInline struct SQRing: ~Copyable {
+    let kernelHead: UnsafePointer<Atomic<UInt32>>
+    let kernelTail: UnsafePointer<Atomic<UInt32>>
```

Comment:
should there be padding around these to prevent false sharing?

Reply:
Hmm possibly. Let me look over how they're used, but that's a good point to consider.